### PR TITLE
[Feature] add list partition support

### LINKED ISSUE #5402: [Proposal] Support of LIST partition
# List partition

Doris currently only supports Range partitioning, where data is usually partitioned by time columns. 

However, in some scenarios, users want to partition by some enumerated values of columns, such as by city, etc.

# Design 

To add support for List partitioning, the following functional points need to be considered.

1. Support for List partition syntax in creating table statements.
2. Support for adding and deleting List partition syntax. 
3. Support for List partitioning in various load operations.
4. Support for List partition pruning during query.

List partitioned tables do not need to consider dynamic partitioning.

# Detailed design

## Syntax

The main changes involved here include.

1. Implementation of the subclass ListPartitionDesc of the parsing class PartitionDesc
2. Implementation of metadata class PartitionInfo subclass ListPartitionInfo
3. Support for parsing and checking ListPartitionDesc in CreateTableStmt
4. Support for the creation of List Partition tables in Catalog class. 
5. Metadata persistence-related changes.

The syntax is referenced from MySQL and Oracle

### Single partition column 

  ```sql
  CREATE TABLE tb1 (
      k1 int, k2 varchar(128), k3 int, v1 int, v2 int
  )
  PARTITION BY LIST(k1)
  (
      PARTITION p1 VALUES IN ("1", "3", "5"),
      PARTITION p2 VALUES IN ("2", "4", "6"),
      ...
  )
  ...
  ;
  ```

### Multi-partition columns

```sql
CREATE TABLE tb2 (
    k1 int, k2 varchar(128), k3 int, v1 int, v2 int
)
PARTITION BY LIST(k1, k2)
(
    PARTITION p1 VALUES IN (("1", "beijing"), ("1", "shanghai")),
    PARTITION p2 VALUES IN (("2", "beijing"), ("2", "shanghai"), ("2", "tianjin")),
    PARTITION p3 VALUES IN (("3", "beijing")),
    ...
)
...
;
``` 

**NOTE**: Each partition needs to ensure that the partition values are unique.

### Add partition 

```sql
ALTER TABLE tb1 ADD PARTITION p4 VALUES IN ("7", "8", "9");

ALTER TABLE tb2 ADD PARTITION p4 VALUES IN (("4", "tianjin"));
```

## Load

The current load methods of Doris include Stream Load, INSERT, Routine Load, Broker Load, Hadoop Load, Spark Load.

Among them, Stream Load, INSERT, Routine Load, and Broker Load all use TabletSink class for data distribution. Our first phase supports List partition support for these load operations.

The main changes involved include: 

1. Changes related to the Descriptors.TOlapTablePartitionParam structure in the Thrift structure TOlapTableSink
2. Changes related to the OlapTablePartition object in the OlapTableSink class on the BE side.

## Query

The query mainly needs to implement the List Partition pruning function.

The main changes involved include: 

1. Implementing the subclass ListPartitionPruner of PartitionPruner

## Partition related

Support operations related to partitioned tables, such as recover, truncate, temporary partition, restore, replace, etc.

### PR DESCRIPTION
**Pull request by qidaye**
## Proposed changes

add list partition support 

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)
- [x] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #5402) and described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If these changes need document changes, I have updated the document
- [x] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: docs/en/getting-started/data-partition.md [hunk 1]
```diff
@@ -86,6 +88,38 @@ PROPERTIES
     "storage_cooldown_time" = "2018-01-01 12:00:00"
 );
 
+
+-- List Partition
+
+CREATE TABLE IF NOT EXISTS example_db.expamle_list_tbl
+(
+    `user_id` LARGEINT NOT NULL COMMENT "用户id",
```
Comment (EmmyMiao87) [inline docs/en/getting-started/data-partition.md:96]: Change it to English.

File: docs/zh-CN/administrator-guide/load-data/delete-manual.md [hunk 2]
```diff
@@ -66,7 +66,7 @@ DELETE FROM my_table PARTITION p1 WHERE k1 < 3 AND k2 = "abc";
 2. `Where`语句中的列只能是`key`列
 3.  当选定的`key`列不存在某个rollup表内时，无法进行delete
 4.  条件语句中各个条件只能是`and`关系，如希望达成`or`可将条件分别写入两个delete语句中
-5.  如果指定表为RANGE分区表，则必须指定 `PARTITION`。如果是单分区表，可以不指定。
+5.  如果指定表为RANGE或者LIST分区表，则必须指定 `PARTITION`。如果是单分区表，可以不指定。
```
Comment (EmmyMiao87) [inline docs/zh-CN/administrator-guide/load-data/delete-manual.md:69]: 英文词的左右两边可以加上空格，这样好看一些

File: be/src/exec/tablet_info.cpp [hunk 3]
```diff
@@ -237,14 +255,27 @@ Status OlapTablePartitionParam::init() {
...
+            _partitions_map->emplace(part->end_key, part);
+        }
     }
     return Status::OK();
 }
 
 bool OlapTablePartitionParam::find_tablet(Tuple* tuple, const OlapTablePartition** partition,
                                           uint32_t* dist_hashes) const {
-    auto it = _partitions_map->upper_bound(tuple);
+    const TOlapTablePartition& t_part = _t_param.partitions[0];
```
[the remark this answers is not in the document]
Comment (qidaye) [reply be/src/exec/tablet_info.cpp:271]: Doris creates at least one partition (unpartitioned table) for the olap table, so here `_t_param.partitions.size() ` is at least 1, no need to worry about.

File: be/src/exec/tablet_info.cpp [hunk 4]
```diff
@@ -237,14 +255,27 @@ Status OlapTablePartitionParam::init() {
...
+    decltype(_partitions_map->find(tuple)) it;
```
[the remark this answers is not in the document]
Comment (qidaye) [reply be/src/exec/tablet_info.cpp]: done

File: docs/en/getting-started/data-partition.md (hunk 1 above)
Comment (qidaye) [reply docs/en/getting-started/data-partition.md]: done

File: docs/zh-CN/administrator-guide/load-data/delete-manual.md (hunk 2 above)
Comment (qidaye) [reply docs/zh-CN/administrator-guide/load-data/delete-manual.md]: done

**Review by morningman — DISMISSED** [review]
LGTM

**Comment by morningman** [comment]
There is some thing to do with #5667.

**Comment by qidaye** [comment]
> There is some thing to do with #5667.
> Better to suppor it after #5667 being merged.

ok, I'll check it.

**Comment by qidaye** [comment]
> > There is some thing to do with #5667.
> > Better to suppor it after #5667 being merged.
> 
> ok, I'll check it.

done

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**